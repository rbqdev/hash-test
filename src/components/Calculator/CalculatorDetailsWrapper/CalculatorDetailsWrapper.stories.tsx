import { Story, Meta } from "@storybook/react/types-6-0";
import CalculatorDetailsWrapper from ".";

export default {
  title: "CalculatorDetailsWrapper",
  component: CalculatorDetailsWrapper
} as Meta;

export const Default: Story = args => <CalculatorDetailsWrapper {...args} />;
