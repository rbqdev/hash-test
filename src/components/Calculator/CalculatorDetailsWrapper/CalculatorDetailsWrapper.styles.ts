import styled, { css } from "styled-components";
import { lighten } from "polished";
import { _constants as themeConstants } from "@styles/themes";
import media from "styled-media-query";

const { THEME_LIGHT } = themeConstants;

export const Wrapper = styled.article`
  display: flex;

  ${media.greaterThan("medium")`
    width: 100%;
  `}
`;

export const WrapperArticle = styled.div`
  ${({ theme: { title, colors } }) => css`
    background: ${title === THEME_LIGHT
      ? colors.secondaryBg
      : lighten(0.04, colors.secondaryBg)};
    padding: 40px;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    font-style: italic;

    ${media.greaterThan("medium")`
      border-radius: 0 4px 4px 0;
    `}

    ${media.lessThan("medium")`
      width: 100%;
      height: auto;
      position: absolute;
      bottom: -100%;
      left: 0;
      visibility: hidden;
      z-index: 2;
      transition: all 0.2s;
    `}
  `}
`;

export const Title = styled.h2`
  text-transform: uppercase;
  border-bottom: 1px solid
    ${({ theme: { title, colors } }) =>
      title === THEME_LIGHT ? colors.blueBorderLight : colors.border};
  color: ${({ theme }) => theme.colors.blue};
  font-size: 16px;
  font-weight: bold;
  padding-bottom: 4px;
  display: block;
  margin-bottom: 15px;
`;
