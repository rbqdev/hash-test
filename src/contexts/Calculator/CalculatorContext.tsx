import React, { createContext } from "react";
import useFetch, { IInitialState } from "@hooks/useFetch/useFetch";
import getUrlAnticipation from "@utils/getUrlAnticipation";

interface IRequestBody {
  amount: number;
  installments: number;
  mdr: number;
  days: number[];
}

interface ICalculatorContext extends IInitialState {
  requestAnticipation: (payload: IRequestBody) => void;
  resetFetchValues: () => void;
}

const CalculatorContext = createContext({} as ICalculatorContext);

export const CalculatorProvider: React.FunctionComponent = ({ children }) => {
  const {
    response,
    error,
    isLoading,
    requestTakeToLong,
    dispatchRequest,
    resetFetchValues
  } = useFetch();

  const requestAnticipation = async (payload: IRequestBody): Promise<void> => {
    const url = getUrlAnticipation();

    await dispatchRequest({
      url,
      method: "POST",
      body: payload
    });
  };

  return (
    <CalculatorContext.Provider
      value={{
        response,
        error,
        isLoading,
        requestTakeToLong,
        requestAnticipation,
        resetFetchValues
      }}
    >
      {children}
    </CalculatorContext.Provider>
  );
};

export default CalculatorContext;
