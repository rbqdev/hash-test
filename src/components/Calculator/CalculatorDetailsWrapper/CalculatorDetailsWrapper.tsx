import { useContext, useState, useEffect } from "react";
import CalculatorContext from "@contexts/Calculator";
import CalculatorDetailsValues from "@components/Calculator/CalculatorDetailsValues";
import * as Styled from "./CalculatorDetailsWrapper.styles";

export default function CalculatorDetailsWrapper() {
  const { response, resetFetchValues, isLoading } = useContext(
    CalculatorContext
  );

  const [isMobileVisible, setIsMobileVisible] = useState(false);

  useEffect(() => {
    const notHasValues = !response || Object.keys(response).length === 0;

    if (notHasValues) {
      return;
    }

    setIsMobileVisible(true);
  }, [response]);

  const handleOverlayClick = () => {
    setIsMobileVisible(false);
    resetFetchValues();
  };

  return (
    <Styled.Wrapper>
      <Styled.WrapperArticle isMobileVisible={isMobileVisible}>
        <Styled.Title>Você Receberá:</Styled.Title>

        <CalculatorDetailsValues
          values={response ?? {}}
          isLoading={isLoading}
        />
      </Styled.WrapperArticle>

      <Styled.OverlayMobile
        onClick={handleOverlayClick}
        isMobileVisible={isMobileVisible}
        aria-hidden={!isMobileVisible}
      />
    </Styled.Wrapper>
  );
}
