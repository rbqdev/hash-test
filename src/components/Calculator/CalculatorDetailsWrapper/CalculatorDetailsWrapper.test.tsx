import CalculatorContext from "@contexts/Calculator";
import render from "@testsUtils/renderWithTheme";
import { fireEvent } from "@testing-library/react";
import CalculatorDetailsWrapper from ".";

const mobileOverlayId = "overlay-mobile";
const mockResetFetchValues = jest.fn();
const mockContextValues = {
  response: null,
  error: null,
  isLoading: false,
  requestTakeToLong: false,
  requestAnticipation: () => {},
  resetFetchValues: mockResetFetchValues
};

const renderComponent = () =>
  render(
    <CalculatorContext.Provider
      value={{
        ...mockContextValues
      }}
    >
      <CalculatorDetailsWrapper />
    </CalculatorContext.Provider>
  );

describe("Component: CalculatorDetailsWrapper", () => {
  it("should render component correctly", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render title", () => {
    const { getByRole } = renderComponent();

    expect(
      getByRole("heading", { name: /você receberá/i })
    ).toBeInTheDocument();
  });

  it("should call 'resetFetchValues' when clicked on overlay", () => {
    const { getByTestId } = renderComponent();

    fireEvent.click(getByTestId(mobileOverlayId));

    expect(mockResetFetchValues).toBeCalledTimes(1);
  });
});
