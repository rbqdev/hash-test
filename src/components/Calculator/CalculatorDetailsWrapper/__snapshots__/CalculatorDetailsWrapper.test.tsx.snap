// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalculatorDetailsWrapper should render component correctly 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
}

.c4 {
  color: #5d9cec;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 46px;
}

.c4 > * {
  white-space: nowrap;
}

.c4 h3 {
  font-weight: normal;
  font-size: 1rem;
  margin: 0;
}

.c4 strong {
  margin-left: 3px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: auto;
  width: 100%;
}

.c1 {
  background: #f7f9fa;
  padding: 40px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-style: italic;
  border-radius: 0 4px 4px 0;
}

.c2 {
  text-transform: uppercase;
  border-bottom: 1px solid #dfecf8;
  color: #3d75bb;
  font-size: 16px;
  font-weight: bold;
  padding-bottom: 4px;
  display: block;
  margin-bottom: 15px;
}

@media (max-width:768px) {
  .c0 {
    width: auto;
  }
}

@media (max-width:768px) {
  .c1 {
    width: 100%;
    height: auto;
    position: absolute;
    bottom: -100%;
    left: 0;
    visibility: hidden;
    z-index: 2;
    -webkit-transition: all 0.2s;
    transition: all 0.2s;
    border-radius: 0;
  }
}

@media (max-width:768px) {
  .c5 {
    width: 100%;
    height: 100%;
    background: #00000099;
    position: fixed;
    top: 0;
    left: 0;
    visibility: hidden;
    opacity: 0;
  }
}

<body>
  <div>
    <article
      class="c0"
    >
      <div
        class="c1"
      >
        <h2
          class="c2"
        >
          Você Receberá:
        </h2>
        <ul
          class="c3"
        >
          <li
            class="c4"
          >
            <h3>
              Amanhã
              :
            </h3>
            <strong>
              R$ 0,00
            </strong>
          </li>
          <li
            class="c4"
          >
            <h3>
              Em 15 dias
              :
            </h3>
            <strong>
              R$ 0,00
            </strong>
          </li>
          <li
            class="c4"
          >
            <h3>
              Em 30 dias
              :
            </h3>
            <strong>
              R$ 0,00
            </strong>
          </li>
          <li
            class="c4"
          >
            <h3>
              Em 90 dias
              :
            </h3>
            <strong>
              R$ 0,00
            </strong>
          </li>
        </ul>
      </div>
      <div
        aria-hidden="true"
        class="c5"
        data-testid="overlay-mobile"
      />
    </article>
  </div>
</body>
`;
